{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output dir: \u001b[34m/home/pesho/repos/bioviz/out/20171121_19h54m01s_batches.WaGSE+WaGSL+WaGBE+WaGBL+WaGBL-C+SchHSE+SchHSL+SchHBE+SchHBL+FrKSE+FrKSL+FrKBE+FrKBL+fluidigm+batch1_DE.wilcox_filt.(cells_in_well1)_experiment.heatmap\u001b[0m\n",
      "\u001b[38;5;220m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;220mWaG_SE\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;214m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;214mWaG_SL\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;216m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;216mWaG_BE\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;27m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;27mWaG_BL\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;138m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;138mWaG_BL-C\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;178m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;178mSchH_SE\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;166m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;166mSchH_SL\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;176m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;176mSchH_BE\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;91m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;91mSchH_BL\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;154m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;154mFrK_SE\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;34m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;34mFrK_SL\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;203m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;203mFrK_BE\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;88m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;88mFrK_BL\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;26m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;26mfluidigm\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\u001b[38;5;226m -----------------------------\u001b[0m\n",
      "Loading batch \u001b[38;5;226mbatch1\u001b[0m...\n",
      "Log2 transformed expressions!\n",
      "Selecting the cells by: \u001b[34m(cells_in_well == 1)\u001b[0m\n",
      "\n",
      "\u001b[38;5;220mWaG_SE\u001b[0m: \u001b[34m84\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_WaG_SE.png\u001b[0m\n",
      " --= batch WaG_SE =--\n",
      "    wells (\u001b[34m84\u001b[0m)\n",
      "    samples: 'SE'(\u001b[34m84\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m76\u001b[0m), '1c'(\u001b[34m6\u001b[0m), '1d'(\u001b[34m2\u001b[0m)\n",
      "\u001b[34m84\u001b[0m cells by \u001b[34m92739\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;214mWaG_SL\u001b[0m: \u001b[34m80\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_WaG_SL.png\u001b[0m\n",
      " --= batch WaG_SL =--\n",
      "    wells (\u001b[34m80\u001b[0m)\n",
      "    samples: 'SL'(\u001b[34m80\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m68\u001b[0m), '1d'(\u001b[34m11\u001b[0m), '1or2'(\u001b[34m1\u001b[0m)\n",
      "\u001b[34m80\u001b[0m cells by \u001b[34m97705\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;216mWaG_BE\u001b[0m: \u001b[34m83\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_WaG_BE.png\u001b[0m\n",
      " --= batch WaG_BE =--\n",
      "    wells (\u001b[34m83\u001b[0m)\n",
      "    samples: 'BE'(\u001b[34m83\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m70\u001b[0m), '1d'(\u001b[34m13\u001b[0m)\n",
      "\u001b[34m83\u001b[0m cells by \u001b[34m76849\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;27mWaG_BL\u001b[0m: \u001b[34m80\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_WaG_BL.png\u001b[0m\n",
      " --= batch WaG_BL =--\n",
      "    wells (\u001b[34m80\u001b[0m)\n",
      "    samples: 'BL'(\u001b[34m80\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m70\u001b[0m), '1c'(\u001b[34m4\u001b[0m), '1d'(\u001b[34m6\u001b[0m)\n",
      "\u001b[34m80\u001b[0m cells by \u001b[34m83363\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;138mWaG_BL-C\u001b[0m: \u001b[34m74\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_WaG_BL-C.png\u001b[0m\n",
      " --= batch WaG_BL-C =--\n",
      "    wells (\u001b[34m74\u001b[0m)\n",
      "    samples: 'BL-C'(\u001b[34m74\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m74\u001b[0m)\n",
      "\u001b[34m74\u001b[0m cells by \u001b[34m97913\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;178mSchH_SE\u001b[0m: \u001b[34m58\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_SchH_SE.png\u001b[0m\n",
      " --= batch SchH_SE =--\n",
      "    wells (\u001b[34m58\u001b[0m)\n",
      "    samples: 'SE'(\u001b[34m58\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m58\u001b[0m)\n",
      "\u001b[34m58\u001b[0m cells by \u001b[34m98719\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;166mSchH_SL\u001b[0m: \u001b[34m53\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_SchH_SL.png\u001b[0m\n",
      " --= batch SchH_SL =--\n",
      "    wells (\u001b[34m53\u001b[0m)\n",
      "    samples: 'SL'(\u001b[34m53\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m53\u001b[0m)\n",
      "\u001b[34m53\u001b[0m cells by \u001b[34m72115\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;176mSchH_BE\u001b[0m: \u001b[34m65\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_SchH_BE.png\u001b[0m\n",
      " --= batch SchH_BE =--\n",
      "    wells (\u001b[34m65\u001b[0m)\n",
      "    samples: 'BE'(\u001b[34m65\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m65\u001b[0m)\n",
      "\u001b[34m65\u001b[0m cells by \u001b[34m55346\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;91mSchH_BL\u001b[0m: \u001b[34m84\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_SchH_BL.png\u001b[0m\n",
      " --= batch SchH_BL =--\n",
      "    wells (\u001b[34m84\u001b[0m)\n",
      "    samples: 'BL'(\u001b[34m84\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m84\u001b[0m)\n",
      "\u001b[34m84\u001b[0m cells by \u001b[34m49059\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;154mFrK_SE\u001b[0m: \u001b[34m74\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_FrK_SE.png\u001b[0m\n",
      " --= batch FrK_SE =--\n",
      "    wells (\u001b[34m74\u001b[0m)\n",
      "    samples: 'SE'(\u001b[34m74\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m74\u001b[0m)\n",
      "\u001b[34m74\u001b[0m cells by \u001b[34m116024\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;34mFrK_SL\u001b[0m: \u001b[34m76\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_FrK_SL.png\u001b[0m\n",
      " --= batch FrK_SL =--\n",
      "    wells (\u001b[34m76\u001b[0m)\n",
      "    samples: 'SL'(\u001b[34m76\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m76\u001b[0m)\n",
      "\u001b[34m76\u001b[0m cells by \u001b[34m88429\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;203mFrK_BE\u001b[0m: \u001b[34m86\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_FrK_BE.png\u001b[0m\n",
      " --= batch FrK_BE =--\n",
      "    wells (\u001b[34m86\u001b[0m)\n",
      "    samples: 'BE'(\u001b[34m86\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m80\u001b[0m), '1d'(\u001b[34m6\u001b[0m)\n",
      "\u001b[34m86\u001b[0m cells by \u001b[34m74882\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;88mFrK_BL\u001b[0m: \u001b[34m73\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_FrK_BL.png\u001b[0m\n",
      " --= batch FrK_BL =--\n",
      "    wells (\u001b[34m73\u001b[0m)\n",
      "    samples: 'BL'(\u001b[34m73\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m73\u001b[0m)\n",
      "\u001b[34m73\u001b[0m cells by \u001b[34m80641\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;26mfluidigm\u001b[0m: \u001b[34m249\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_fluidigm.png\u001b[0m\n",
      " --= batch fluidigm =--\n",
      "    wells (\u001b[34m249\u001b[0m)\n",
      "    samples: '7010'(\u001b[34m72\u001b[0m), '7015'(\u001b[34m63\u001b[0m), '7014'(\u001b[34m62\u001b[0m), '7011'(\u001b[34m52\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m249\u001b[0m)\n",
      "\u001b[34m249\u001b[0m cells by \u001b[34m173259\u001b[0m transcripts\n",
      "\n",
      "\u001b[38;5;226mbatch1\u001b[0m: \u001b[34m14\u001b[0m cells\n",
      "\n",
      "Figure saved as \u001b[34mstats_batch1.png\u001b[0m\n",
      " --= batch batch1 =--\n",
      "    wells (\u001b[34m14\u001b[0m)\n",
      "    samples: 'A'(\u001b[34m14\u001b[0m)\n",
      "    well counts: '1'(\u001b[34m13\u001b[0m), '1d'(\u001b[34m1\u001b[0m)\n",
      "\u001b[34m13\u001b[0m cells by \u001b[34m173259\u001b[0m transcripts\n",
      "\n",
      "\u001b[31m   ------- [PDF] --------\u001b[0m\n",
      "Make a copy of the notebook...\n",
      "Collect \u001b[34m15\u001b[0m images and stamp them...\n",
      "Merge all images to one PDF [\u001b[34mstats_20171121_19h54m01s_batches.WaGSE+WaGSL+WaGBE+WaGBL+WaGBL-C+SchHSE+SchHSL+SchHBE+SchHBL+FrKSE+FrKSL+FrKBE+FrKBL+fluidigm+batch1_DE.wilcox_filt.(cells_in_well1).pdf\u001b[0m]...\n",
      "\u001b[31m   ------- [DONE] --------\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "#%%prun\n",
    "####  THIS CELL CLEARS THE STATE  ####\n",
    "#%load_ext line_profiler\n",
    "#%install_ext https://raw.github.com/cpcloud/ipython-autotime/master/autotime.py\n",
    "%reset -f\n",
    "#%reload_ext autotime\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib notebook\n",
    "#%load_ext memory_profiler\n",
    "\n",
    "from imports import *\n",
    "run.init(samples = [\n",
    "    ### WTF: TAGA(G)GACT should be equal to TAGA(G)GACT\n",
    "    ('WaG', 'SE',   'sunflower yellow', 'TRBV20-1_CTAGATCGAGACTAGCGGGCTCCT_TRBJ2-7'), # orig: 'TRBV20-1_CTAGA(G)GACTAGCGGGCTCCT_TRBJ2-1'\n",
    "    ('WaG', 'SL',   'goldenrod',        'TRBV20-1_CTAGATCGAGACTAGCGGGCTCCT_TRBJ2-7'), #'TRAV26-1_AGTCGTTAGTG_TRAJ53 or TRBV20-1_CTAGA(G)GACTAGCGGGCTCCT_TRBJ2-1'), #NO good clones#\n",
    "    ('WaG', 'BE',   'peach',            'TRBV20-1_CTAGATCGAGACTAGCGGGCTCCT_TRBJ2-7'), #'TRAV10_GTGAGGGACCCCATCCTCTG_TRAJ28 or TRBV6-5_GTTACGGGGACAAGTACGA_TRBJ2-7'), #NO good clones#\n",
    "    ('WaG', 'BL',   'bright blue',      'TRBV20-1_CTAGATCGAGACTAGCGGGCTCCT_TRBJ2-7'), #'TRAV35_GGCAGGGGTCAAGTGC_TRAJ3 or TRBV6-5_GTTACGGGGACAAGTACGA_TRBJ2-7'),\n",
    "    ('WaG', 'BL-C', 'dusty rose',       'TRBV20-1_CTAGATCGAGACTAGCGGGCTCCT_TRBJ2-7'),  # 11 cells\n",
    "###    ('WaG', 'BL-C', 'azure',            'TRAV9-2_CTTTG(A)ACTTT_TRAJ48'),  # 22 cells\n",
    "\n",
    "    ('SchH', 'SE', 'mustard yellow',  'TRBV20-1_AGAGATTTAACTAGCGGGAGCTTCCCCTAC_TRBJ2-7'), #'TRAV12-2_TGAACTGGGATA_TRAJ33 or TRBV20-1_AGAGATTTAACTAGCGGGAGCTTCCCCTAC_TRBJ2-7'),\n",
    "    ('SchH', 'SL', 'dark orange',     'TRBV20-1_AGAGATTTAACTAGCGGGAGCTTCCCCTAC_TRBJ2-7'),\n",
    "    ('SchH', 'BE', 'orchid',          'TRBV20-1_AGAGATTTAACTAGCGGGAGCTTCCCCTAC_TRBJ2-7'),  # old hypo: 'TRAV13-1_AGCAACCTTTCCTAAC_TRAJ20'\n",
    "    ('SchH', 'BL', 'purple',          'TRBV20-1_AGAGATTTAACTAGCGGGAGCTTCCCCTAC_TRBJ2-7'), #NO good clones#\n",
    "    \n",
    "    ('FrK', 'SE', 'chartreuse', 'TRBV4-1_CCAAGCCCTCGGGGGGGATAAT_TRBJ1-6'),\n",
    "    ('FrK', 'SL', 'green',      'TRBV4-1_CCAAGCCCTCGGGGGGGATAAT_TRBJ1-6'), #'TRBV4-1_CCAAGCCCTCGGGGGGGATAAT_TRBJ1-6'),\n",
    "    ('FrK', 'BE', 'coral',      'TRBV4-1_CCAAGCCCTCGGGGGGGATAAT_TRBJ1-6'), #NO good clones#\n",
    "    ('FrK', 'BL', 'crimson',    'TRBV4-1_CCAAGCCCTCGGGGGGGATAAT_TRBJ1-6'),  # TRAV6_TGTGCCGTACGGGATA_TRAJ12 or TRBV4-1_CCAAGCCCTCGGGGGGGATAAT_TRBJ1-6'\n",
    "\n",
    "    ('fluidigm', '', 'blue',    'NoClone'),\n",
    "    ('batch1', '', 'yellow',    'NoClone'),\n",
    "],\n",
    "    \n",
    "    experiment = 'heatmap',\n",
    "    notebook = 'bioviz.ipynb',\n",
    "    input_files = { 'expression_matrix': \"expressions.tsv\" },\n",
    "    nrows = 0,\n",
    "    #features = [10, 'important', 'all'],\n",
    "    features = [ 'all' ],\n",
    "    diffexprs = ['wilcox'], # 'ks'\n",
    "    interactive = False,\n",
    "    run = {\n",
    "           'stats': True,\n",
    "           #'vb_distr':      1\n",
    "           'filt_cells': {\n",
    "               'cells': '(cells_in_well == 1)',\n",
    "               'expressed_genes': 1500,\n",
    "           },\n",
    "           'filt_genes': {\n",
    "               'expression_threshold': 2.5,\n",
    "               'min_cells_expressing_gene': 15,\n",
    "           },\n",
    "           #'filt': {'cells': '(cells_in_well == 1) & (mapped > 500K)'},\n",
    "           #'normalize': 'byTotalSum', # 'bySpikeIn'\n",
    "           #'combine_batches': ['patient'],   # ['patient', 'pairs', 'all'],\n",
    "           #'scatter': { 'dim_reduct': ['PCA'],     # 't-SNE', 'MDS', \"ZIFA\"\n",
    "           #             'coloring': ['byClone'],\n",
    "           #             'coloring': ['bySample', 'byClone', 'bySampleAndClone'],\n",
    "           #           },\n",
    "           #'classifier':    0,\n",
    "        \n",
    "        #   'geneintersect': True,\n",
    "           #'quantheatmap':  {'coloring': ['byClone']},\n",
    "        \n",
    "           #'quantheatmap':  {'coloring': ['bySample', 'byClone', 'bySampleAndClone']},\n",
    "           #'bipartite':     ['clonal', 'sample', 'rainbow'],\n",
    "           #'bipartite':     ['clonal'],      # ['clonal', 'sample', 'rainbow'],\n",
    "#           'gsea': {'geneset': 'human_all_collections_lymphocytes_sept2017.gmt'},\n",
    "           #'gsea': {'geneset': 'c1.all.v6.0.symbols.gmt'}, # 'gseaftp.broadinstitute.org://pub/gsea/gene_sets_final/c5.cc.v3.1.symbols.gmt'},\n",
    "    },\n",
    "    kernels = ['decision-tree', 'log-linear', 'rbf'],   # ['linear', 'lin', 'poly', 'rbf'],\n",
    "    Cs = [1], gammas = [1],\n",
    "    #Cs = [1e-10, 1e-5, 1e0, 1e5, 1e10], gammas = [1e-10, 1e-5, 1e0, 1e5, 1e10] \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from algo.normalize import NormalizeByTotalSum\n",
    "df = pd.DataFrame({'a': [4, 5, 6], 'b': [47, 55, 26], })\n",
    "display(df)\n",
    "NormalizeByTotalSum(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import rpy2.robjects as robj\n",
    "from rpy2.robjects import numpy2ri\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects.packages import importr\n",
    "\n",
    "from utils import *\n",
    "\n",
    "stats = importr(\"stats\")\n",
    "gplots = importr(\"gplots\")\n",
    "\n",
    "numpy2ri.activate()\n",
    "pandas2ri.activate()\n",
    "R = robj.r\n",
    "rprint = robj.globalenv.get(\"print\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import rpy2.robjects as robj\n",
    "from rpy2.robjects import numpy2ri\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects.packages import importr\n",
    "\n",
    "deseq = importr(\"DESeq\")\n",
    "numpy2ri.activate()\n",
    "pandas2ri.activate()\n",
    "\n",
    "#datafile = R.system.file( \"extdata/pasilla_gene_counts.tsv\", package=\"pasilla\" )\n",
    "#d = {'a': robj.IntVector((11,12,13)), 'b': robj.IntVector((4,5,6))}\n",
    "def normalize(countTable):\n",
    "    conditions = robj.FactorVector( 'c' * len(countTable.columns) )\n",
    "    cds = deseq.newCountDataSet( countTable, conditions )\n",
    "    data = cds.slots['assayData']['counts']\n",
    "    print(data)\n",
    "    sizeFactors = deseq.estimateSizeFactorsForMatrix( data )\n",
    "    return countTable / list(sizeFactors)\n",
    "countTable = pd.DataFrame({'a': [11, 12, 13], 'b': [4, 5, 6]})\n",
    "display(countTable)\n",
    "display(normalize(countTable))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "list(sizeFactors)\n",
    "#countTable.divide(sizeFactors, axis='columns')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#b=run.B[('FrK', 'SE')]\n",
    "for k, b in run.B.items():\n",
    "    print(b)\n",
    "    print([t for t in b.quant.transcripts() if not t.startswith('ENST') ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import Javascript\n",
    "from IPython.display import display\n",
    "import time\n",
    "\n",
    "def get_notebook_name():\n",
    "    \"\"\"Returns the name of the current notebook as a string\n",
    "    \n",
    "    From From https://mail.scipy.org/pipermail/ipython-dev/2014-June/014096.html\n",
    "    \"\"\"\n",
    "    display(Javascript('IPython.notebook.kernel.execute(\"theNotebook = \" + \\\n",
    "    \"\\'\"+IPython.notebook.notebook_name+\"\\'\");'))\n",
    "\n",
    "    time.sleep(1)\n",
    "    return theNotebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "display(\"text/html\",\n",
    "\"\"\"\n",
    "<script charset=\"utf-8\">\n",
    "    julia_code = 'notebook_name = \"'+IPython.notebook.notebook_name+'\";';\n",
    "    IPython.notebook.kernel.execute(julia_code);\n",
    "</script>\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from imports import *\n",
    "import nbformat\n",
    "import nbconvert\n",
    "\n",
    "#print(notebook_name)\n",
    "#print(get_notebook_name())\n",
    "\n",
    "#IPython.notebook.notebook_name\n",
    "\n",
    "nb_name = 'bioviz.ipynb'\n",
    "with open(dirs['src'] / nb_name) as f:\n",
    "    nb = nbformat.read(f, as_version=4)\n",
    "    f.close()\n",
    "    \n",
    "#pdf_exporter = nbconvert.PDFExporter()\n",
    "pdf_exporter = nbconvert.MarkdownExporter()\n",
    "pdf_data, resources = pdf_exporter.from_notebook_node(nb)\n",
    "\n",
    "with open(\"executed_notebook.md\", \"wb\") as f:\n",
    "    f.write(pdf_data)\n",
    "    f.close()\n",
    "    \n",
    "#exp = nbconvert.exporters.PDFExporter\n",
    "#exp = nbconvert.exporters.MarkdownExporter\n",
    "#md = nbconvert.exporters.export(exp, nb)\n",
    "#with open('executed_notebook.md', 'wt') as f:\n",
    "#    f.write(md)\n",
    "#    #nbformat.write(nb, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#import vis.ge_scatter_tcr_graph\n",
    "#%lprun -f vis.ge_scatter_tcr_graph.ScatterTCR.scatter2D run.produce_results()\n",
    "from structs.clone_struct import Hypothesis\n",
    "%lprun -f Hypothesis.calc_log2fc run.B[('FrK','SE')].get_important_genes()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TCR alpha constant (TRAC):\n",
    "Gene: ENSG00000277734\n",
    "Transcript (Yun-Tsan): ENST00000611116\n",
    "['ENST00000611116', 'ENST00000637010', 'ENST00000636588', 'ENST00000616778', 'ENST00000636320']\n",
    "\n",
    "TCR beta constant (TRBC2)\n",
    "Gene:  ENSG00000211772\n",
    "Transcript (Yun-Tsan): ENST00000466254\n",
    "['ENST00000466254', 'ENST00000636844', 'ENST00000610416', 'ENST00000637077', 'ENST00000622053', 'ENST00000614992', 'ENST00000620987', 'ENST00000613720']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#print(run.gcv.run_classifier(kernel='rbf', c=1, gamma=1))\n",
    "#print(run.gcv.params2res)\n",
    "#print(run.gcv.results2html())\n",
    "#sample = run.B['FrK_SL']\n",
    "#hypo = sample.clones.best_hypo()\n",
    "#print(hypo)\n",
    "#pprint(hypo.best_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# +  1) lower the number of transcripts (clean the zeros)\n",
    "# +  2) unite hypotheses, subcell hypotheses (with old p-values!)\n",
    "#    3) qc filtering (incl. tcr len)\n",
    "# +  4) dimreduct over important genes\n",
    "# +  5) lighter colors for non-clonal\n",
    "#    6) violin plots for important genes\n",
    "# +  7) clonal definitions: single chain, union, intersection, non-clonal definition\n",
    "# +  7.1) autogenerate clonal combinations\n",
    "# +  8) classifier (SVM)\n",
    "# ?  9) cross-validation (honest)\n",
    "# + 10) gene list union\n",
    "# ? 11) TCR quantification and bio validation\n",
    "# + 12) diff expr: ks-test, wilcox\n",
    "#   13) diff expr: single-cell specific\n",
    "#   14) ipywidgets: expose web interface\n",
    "#   15) parallelize gene testing\n",
    "#   16) compare clones with each other\n",
    "\n",
    "#   * (C x kernel) x (patient x sample)\n",
    "#   * groundtruth: K random features, all features"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  },
  "widgets": {
   "state": {
    "4b291cd54c6a42d3abc392eb90e8f995": {
     "views": [
      {
       "cell_index": 5
      }
     ]
    }
   },
   "version": "1.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
